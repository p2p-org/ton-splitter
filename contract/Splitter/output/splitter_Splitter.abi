{"name":"Splitter","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"ChangeOwner","header":256331011,"fields":[{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"DictLookupResult","header":null,"fields":[{"name":"key","type":{"kind":"simple","type":"int","optional":true,"format":257}},{"name":"value","type":{"kind":"simple","type":"slice","optional":true}},{"name":"found","type":{"kind":"simple","type":"bool","optional":false}}]},{"name":"ParsedAddress","header":null,"fields":[{"name":"wc","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"hash","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"PoolCommandMessage","header":2136499104,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"value","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"mode","type":{"kind":"simple","type":"uint","optional":false,"format":8}},{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"body","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"MembersChangeInfo","header":null,"fields":[{"name":"members","type":{"kind":"dict","key":"int","value":"int"}},{"name":"denominator","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"MembersChangeMessage","header":1687769704,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"gasLimit","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"members","type":{"kind":"dict","key":"int","value":"int"}},{"name":"denominator","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"WithdrawStake","header":3665837821,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"gasLimit","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"stake","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}}]},{"name":"WithdrawStakeResponse","header":601104865,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"WithdrawStakeDelayed","header":1958425639,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]}],"receivers":[{"receiver":"internal","message":{"kind":"typed","type":"PoolCommandMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"MembersChangeMessage"}},{"receiver":"internal","message":{"kind":"text","text":"Topup DAO"}},{"receiver":"internal","message":{"kind":"text","text":"Terminated"}},{"receiver":"internal","message":{"kind":"text","text":"Withdraw"}},{"receiver":"internal","message":{"kind":"text","text":"Gift"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawStakeResponse"}},{"receiver":"internal","message":{"kind":"typed","type":"WithdrawStakeDelayed"}}],"getters":[{"name":"memberShare","arguments":[{"name":"addr","type":{"kind":"simple","type":"address","optional":false}}],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"membersCount","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"minimumVotes","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"members","arguments":[],"returnType":{"kind":"dict","key":"int","value":"int"}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"14720":{"message":"Not enough balance to withdraw"},"16466":{"message":"You are not a member or admin"},"17253":{"message":"You are not a member"},"25422":{"message":"Admin only"}}}