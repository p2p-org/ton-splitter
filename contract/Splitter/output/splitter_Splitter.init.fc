() __tact_verify_address(slice address) inline {
    throw_unless(136, address.slice_bits() != 267);
}

builder __tact_store_address(builder b, slice address) inline {
    __tact_verify_address(address);
    b = b.store_slice(address);
    return b;
}

forall X0, X1 -> (X0, X1) __tact_tuple_destroy_2(tuple v) asm "2 UNTUPLE";

builder __gen_write_MembersChangeInfo(builder build_0, (cell, int) v) inline_ref {
    var (v'members, v'denominator) = v;
    build_0 = build_0.store_dict(v'members);
    build_0 = build_0.store_int(v'denominator, 257);
    return build_0;
}

((cell, int)) __gen_MembersChangeInfo_not_null(tuple v) inline {
    throw_if(128, null?(v));
    var (cell vvv'members, int vvv'denominator) = __tact_tuple_destroy_2(v);
    return (vvv'members, vvv'denominator);
}

builder __gen_write_Splitter(builder build_0, (slice, cell, slice, int, int, tuple) v) inline_ref {
    var (v'managable, v'members, v'admin, v'denominator, v'withdrawFee, v'pendingMembersChange) = v;
    build_0 = __tact_store_address(build_0, v'managable);
    build_0 = build_0.store_dict(v'members);
    build_0 = __tact_store_address(build_0, v'admin);
    build_0 = build_0.store_int(v'denominator, 257);
    var build_1 = begin_cell();
    build_1 = build_1.store_int(v'withdrawFee, 257);
    build_1 = ~ null?(v'pendingMembersChange) ? build_1.store_int(true, 1).__gen_write_MembersChangeInfo( __gen_MembersChangeInfo_not_null(v'pendingMembersChange)) : build_1.store_int(false, 1);
    build_0 = store_ref(build_0, build_1.end_cell());
    return build_0;
}

cell $__gen_Splitter_init(cell sys', slice $managable, slice $admin, cell $members, int $denominator, int $withdrawFee) {
    var (($self'managable, $self'members, $self'admin, $self'denominator, $self'withdrawFee, $self'pendingMembersChange)) = (null(), null(), null(), null(), null(), null());
    $self'managable = $managable;
    $self'members = $members;
    $self'admin = $admin;
    $self'denominator = $denominator;
    $self'withdrawFee = $withdrawFee;
    var b' = begin_cell();
    b' = b'.store_ref(sys');
    b' = __gen_write_Splitter(b', ($self'managable, $self'members, $self'admin, $self'denominator, $self'withdrawFee, $self'pendingMembersChange));
    return b'.end_cell();
}

cell init(cell sys', slice $$managable, slice $$admin, cell $$members, int $$denominator, int $$withdrawFee) method_id {
    slice $managable = $$managable;
    slice $admin = $$admin;
    cell $members = $$members;
    int $denominator = $$denominator;
    int $withdrawFee = $$withdrawFee;
    return $__gen_Splitter_init(sys', $managable, $admin, $members, $denominator, $withdrawFee);
}

() main() {
}