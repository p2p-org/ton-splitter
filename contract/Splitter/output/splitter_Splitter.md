# TACT Compilation Report
Contract: Splitter
BOC Size: 1973 bytes

# Types
Total Types: 14

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## ChangeOwner
TLB: `change_owner#0f474d03 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{newOwner:address}`

## DictLookupResult
TLB: `_ key:Maybe int257 value:Maybe ^slice found:bool = DictLookupResult`
Signature: `DictLookupResult{key:Maybe int257,value:Maybe ^slice,found:bool}`

## ParsedAddress
TLB: `_ wc:int257 hash:int257 = ParsedAddress`
Signature: `ParsedAddress{wc:int257,hash:int257}`

## PoolCommandMessage
TLB: `pool_command_message#7f5863a0 queryId:uint64 value:coins mode:uint8 bounce:bool body:^cell = PoolCommandMessage`
Signature: `PoolCommandMessage{queryId:uint64,value:coins,mode:uint8,bounce:bool,body:^cell}`

## MembersChangeInfo
TLB: `_ members:dict<int, int> denominator:int257 = MembersChangeInfo`
Signature: `MembersChangeInfo{members:dict<int, int>,denominator:int257}`

## MembersChangeMessage
TLB: `members_change_message#64995268 queryId:uint64 gasLimit:coins members:dict<int, int> denominator:int257 = MembersChangeMessage`
Signature: `MembersChangeMessage{queryId:uint64,gasLimit:coins,members:dict<int, int>,denominator:int257}`

## WithdrawStake
TLB: `withdraw_stake#da803efd queryId:uint64 gasLimit:coins stake:coins = WithdrawStake`
Signature: `WithdrawStake{queryId:uint64,gasLimit:coins,stake:coins}`

## WithdrawStakeResponse
TLB: `withdraw_stake_response#23d421e1 queryId:uint64 = WithdrawStakeResponse`
Signature: `WithdrawStakeResponse{queryId:uint64}`

## WithdrawStakeDelayed
TLB: `withdraw_stake_delayed#74bb3427 queryId:uint64 = WithdrawStakeDelayed`
Signature: `WithdrawStakeDelayed{queryId:uint64}`

# Get Methods
Total Get Methods: 4

## memberShare
Argument: addr

## membersCount

## minimumVotes

## members
