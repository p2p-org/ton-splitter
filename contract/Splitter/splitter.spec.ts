import { Address, beginCell, comment, Dictionary, toNano } from "ton";
import { ContractSystem, Treasure } from "ton-emulator";
import { Splitter, storePoolCommandMessage, storeWithdrawStake, WithdrawStake } from "./output/splitter_Splitter";
import assert from "assert";

describe("Splitter", () => {
    it("should split revenue correctly", async () => {
        // Create ContractSystem and deploy contract
        const system = await ContractSystem.create();

        const managable = system.treasure("managable");
        const admin = system.treasure("admin");
        const owner = system.treasure("owner");
        const nonOwner = system.treasure("non-owner");

        const members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        members.set(BigInt("0x" + owner.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + nonOwner.address.hash.toString("hex")), 1n);

        const contract = system.open(
            await Splitter.fromInit(managable.address, admin.address, members, 2n, 200000000n)
        );
        const track = system.track(contract.address);
        await system.run();

        track.events();

        // Check splitter
        await contract.send(
            managable,
            {
                value: toNano(2) + 100_000_000n,
            },
            "Gift"
        );

        await system.run();
        expect(track.events()).toMatchSnapshot();
    });

    it("should split if withdraw delayed", async () => {
        let system = await ContractSystem.create();

        let managable = system.treasure("managable");
        const admin = system.treasure("admin");
        let owner = system.treasure("owner");
        let nonOwner = system.treasure("non-owner");
        let nonOwner2 = system.treasure("non-owner 2");

        let members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        members.set(BigInt("0x" + owner.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + nonOwner.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + nonOwner2.address.hash.toString("hex")), 1n);

        let contract = system.open(await Splitter.fromInit(managable.address, admin.address, members, 3n, 200000000n));
        let track = system.track(contract.address);

        await system.run();
        track.events();

        const amount = toNano(10);
        await contract.send(
            managable,
            {
                value: amount,
                bounce: false,
            },
            {
                $$type: "WithdrawStakeDelayed",
                queryId: 228n,
            }
        );

        await system.run();
        expect(track.events()).toMatchSnapshot();
    });

    it("should split if zero balance", async () => {
        let system = await ContractSystem.create();

        let managable = system.treasure("managable");
        const admin = system.treasure("admin");
        let owner = system.treasure("owner");
        let nonOwner = system.treasure("non-owner");
        let nonOwner2 = system.treasure("non-owner 2");

        let members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        members.set(BigInt("0x" + owner.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + nonOwner.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + nonOwner2.address.hash.toString("hex")), 1n);

        let contract = system.open(await Splitter.fromInit(managable.address, admin.address, members, 3n, 200000000n));
        let track = system.track(contract.address);

        await system.run();
        track.events();

        const amount = toNano(10);
        await contract.send(
            managable,
            {
                value: amount,
                bounce: false,
            },
            {
                $$type: "WithdrawStakeResponse",
                queryId: 228n,
            }
        );

        await system.run();
        expect(track.events()).toMatchSnapshot();
    });

    it("should execute admin commands", async () => {
        const system = await ContractSystem.create();

        const managable = system.treasure("managable");
        const admin = system.treasure("admin");

        // empty members
        const members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));

        const contract = system.open(
            await Splitter.fromInit(managable.address, admin.address, members, 3n, 200000000n)
        );

        const track = system.track(contract.address);

        await system.run();
        track.events();

        /** Forming the message for pool */
        const dataToSend = beginCell()
            .storeUint(123, 32) // op
            .storeUint(123, 64) // query id
            .storeCoins(toNano(10)) // gas limits
            .storeUint(123321123n, 128) // some payload
            .endCell();

        /** Forming the command for splitter */
        const poolCommandBuilder = beginCell();
        storePoolCommandMessage({
            $$type: "PoolCommandMessage",
            queryId: 123n,
            value: 500_000_000n,
            mode: 0n,
            bounce: true,
            body: dataToSend,
        })(poolCommandBuilder);
        const poolCommandMessage = poolCommandBuilder.endCell();

        // sending the command
        await contract.send(
            admin,
            {
                value: toNano(1),
                bounce: true,
            },
            {
                $$type: "PoolCommandMessage",
                queryId: 123n,
                value: 500_000_000n,
                mode: 0n,
                bounce: true,
                body: dataToSend,
            }
        );

        await system.run();
        const events = track.events();

        /** Splitter received the command */
        const toSplitterFromAdming = events.find(
            (e) =>
                e.type === "received" &&
                e.message.type === "internal" &&
                e.message.from === admin.address.toString({ testOnly: true, bounceable: true }) &&
                e.message.to === contract.address.toString({ testOnly: true, bounceable: true })
        );
        assert(toSplitterFromAdming);
        assert(toSplitterFromAdming.type === "received");
        assert(toSplitterFromAdming.message.type === "internal");
        assert(toSplitterFromAdming.message.body.type === "cell");
        expect(toSplitterFromAdming.message.body.cell).toEqual(poolCommandMessage.toString());

        /** Splitter sends the command's body to pool */
        const toPoolFromSplitter = events.find((e) => e.type === "sent" && e.messages.length === 1);
        assert(toPoolFromSplitter);
        assert(toPoolFromSplitter.type === "sent");
        assert(toPoolFromSplitter.messages.length === 1);
        const message = toPoolFromSplitter.messages[0];
        assert(message.type === "internal");
        expect(message.from).toEqual(contract.address.toString({ testOnly: true, bounceable: true }));
        expect(message.to).toEqual(managable.address.toString({ testOnly: true, bounceable: true }));
        assert(message.body.type === "cell");
        expect(message.body.cell).toEqual(dataToSend.toString());
    });

    it("should change members", async () => {
        const system = await ContractSystem.create();

        const managable = system.treasure("managable");
        const admin = system.treasure("admin");

        /** Initial members */
        const member1 = system.treasure("member1");
        const member2 = system.treasure("member2");
        const member3 = system.treasure("member3");

        /** Members dictionary */
        const members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        members.set(BigInt("0x" + member1.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + member2.address.hash.toString("hex")), 2n);
        members.set(BigInt("0x" + member3.address.hash.toString("hex")), 3n);

        const contract = system.open(
            await Splitter.fromInit(managable.address, admin.address, members, 6n, 200000000n)
        );
        const track = system.track(contract.address);
        const tx = await system.run();
        track.events();

        // send some message to deploy the contract
        await contract.send(member1, { value: toNano(1) }, "Gift");

        await system.run();

        const membersBefore = await contract.getMembers();

        /** Check members before */
        expect(membersBefore.size).toBe(3);
        expect(membersBefore.get(BigInt("0x" + member1.address.hash.toString("hex")))).toBe(1n);
        expect(membersBefore.get(BigInt("0x" + member2.address.hash.toString("hex")))).toBe(2n);
        expect(membersBefore.get(BigInt("0x" + member3.address.hash.toString("hex")))).toBe(3n);

        /** New members */
        const member4 = system.treasure("member4");
        const member5 = system.treasure("member5");
        const member6 = system.treasure("member6");
        const member7 = system.treasure("member7");

        /** New members dictionary */
        const newMembers = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        newMembers.set(BigInt("0x" + member4.address.hash.toString("hex")), 1n);
        newMembers.set(BigInt("0x" + member5.address.hash.toString("hex")), 2n);
        newMembers.set(BigInt("0x" + member6.address.hash.toString("hex")), 3n);
        newMembers.set(BigInt("0x" + member7.address.hash.toString("hex")), 4n);

        /** Send new members */
        await contract.send(
            admin,
            { value: toNano(1) },
            {
                $$type: "MembersChangeMessage",
                members: newMembers,
                queryId: 123n,
                gasLimit: toNano(1),
                denominator: 10n,
            }
        );
        await system.run();

        /** Mock pool withdrawal */
        await contract.send(
            managable,
            { value: toNano(100) },
            {
                $$type: "WithdrawStakeResponse",
                queryId: 123n,
            }
        );
        await system.run();

        const membersAfter = await contract.getMembers();

        /** Check members after */
        expect(membersAfter.size).toBe(4);
        expect(membersAfter.get(BigInt("0x" + member4.address.hash.toString("hex")))).toBe(1n);
        expect(membersAfter.get(BigInt("0x" + member5.address.hash.toString("hex")))).toBe(2n);
        expect(membersAfter.get(BigInt("0x" + member6.address.hash.toString("hex")))).toBe(3n);
        expect(membersAfter.get(BigInt("0x" + member7.address.hash.toString("hex")))).toBe(4n);
    });

    it("should let only members and admin withdraw", async () => {
        const system = await ContractSystem.create();

        const managable = system.treasure("managable");
        const admin = system.treasure("admin");

        /** Initial members */
        const member1 = system.treasure("member1");
        const member2 = system.treasure("member2");
        const member3 = system.treasure("member3");

        /** Members dictionary */
        const members = Dictionary.empty(Dictionary.Keys.BigInt(257), Dictionary.Values.BigInt(257));
        members.set(BigInt("0x" + member1.address.hash.toString("hex")), 1n);
        members.set(BigInt("0x" + member2.address.hash.toString("hex")), 2n);
        members.set(BigInt("0x" + member3.address.hash.toString("hex")), 3n);

        const WITHDRAW_FEE = 200000000n;

        const contract = system.open(
            await Splitter.fromInit(managable.address, admin.address, members, 6n, WITHDRAW_FEE)
        );
        const track = system.track(contract.address);
        const tx = await system.run();
        track.events();

        const randomAddress = system.treasure("random");

        await contract.send(randomAddress, { value: WITHDRAW_FEE + toNano(1) }, "Withdraw");
        await system.run();
        expect(track.events()).toMatchSnapshot();

        await contract.send(member1, { value: WITHDRAW_FEE + toNano(1) }, "Withdraw");
        await system.run();
        expect(track.events()).toMatchSnapshot();

        await contract.send(admin, { value: WITHDRAW_FEE + toNano(1) }, "Withdraw");
        await system.run();
        expect(track.events()).toMatchSnapshot();
    });
});
