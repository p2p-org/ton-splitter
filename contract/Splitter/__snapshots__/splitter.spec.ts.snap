// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter should let only members and admin withdraw 1`] = `
[
  {
    "type": "deploy",
  },
  {
    "message": {
      "body": {
        "text": "Withdraw",
        "type": "text",
      },
      "bounce": true,
      "from": "kQBR0KWyao8i8wEuP4v5RguceCxMzKfA50OXlrnZKr4-4yrY",
      "to": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
      "type": "internal",
      "value": 1200000000n,
    },
    "type": "received",
  },
  {
    "errorCode": 16466,
    "type": "failed",
  },
  {
    "message": {
      "body": {
        "type": "empty",
      },
      "bounce": false,
      "from": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
      "to": "kQBR0KWyao8i8wEuP4v5RguceCxMzKfA50OXlrnZKr4-4yrY",
      "type": "internal",
      "value": 1192762000n,
    },
    "type": "sent-bounced",
  },
]
`;

exports[`Splitter should let only members and admin withdraw 2`] = `
[
  {
    "message": {
      "body": {
        "text": "Withdraw",
        "type": "text",
      },
      "bounce": true,
      "from": "kQAddWGzDY_nU13IVl-JpWBXFuB0fGk8wV8I75kLpljvMqxE",
      "to": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
      "type": "internal",
      "value": 1200000000n,
    },
    "type": "received",
  },
  {
    "gasUsed": 13372n,
    "type": "processed",
  },
  {
    "messages": [
      {
        "body": {
          "cell": "x{DA803EFD4E848C973758AB7043B9ACA000}",
          "type": "cell",
        },
        "bounce": false,
        "from": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
        "to": "kQC3OBC-eWfUrnx1Yp6h6sOj779mNY3u3bK4idWfAdFVe7sB",
        "type": "internal",
        "value": 200000000n,
      },
    ],
    "type": "sent",
  },
]
`;

exports[`Splitter should let only members and admin withdraw 3`] = `
[
  {
    "message": {
      "body": {
        "text": "Withdraw",
        "type": "text",
      },
      "bounce": true,
      "from": "kQCSi8fIJH0YiDmgUQ7UTYGIeiLjXq0NN3oDDdznGOn3I03k",
      "to": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
      "type": "internal",
      "value": 1200000000n,
    },
    "type": "received",
  },
  {
    "gasUsed": 13446n,
    "type": "processed",
  },
  {
    "messages": [
      {
        "body": {
          "cell": "x{DA803EFD9225D108166BD8CC43B9ACA000}",
          "type": "cell",
        },
        "bounce": false,
        "from": "kQD_nbFCT5l_uffqhOr1BnXqjSVNgpgW8Ga7OTauiS0C2U9V",
        "to": "kQC3OBC-eWfUrnx1Yp6h6sOj779mNY3u3bK4idWfAdFVe7sB",
        "type": "internal",
        "value": 200000000n,
      },
    ],
    "type": "sent",
  },
]
`;

exports[`Splitter should split if withdraw delayed 1`] = `
[
  {
    "type": "deploy",
  },
  {
    "message": {
      "body": {
        "cell": "x{74BB342700000000000000E4}",
        "type": "cell",
      },
      "bounce": false,
      "from": "kQC3OBC-eWfUrnx1Yp6h6sOj779mNY3u3bK4idWfAdFVe7sB",
      "to": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
      "type": "internal",
      "value": 10000000000n,
    },
    "type": "received",
  },
  {
    "gasUsed": 77532n,
    "type": "processed",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQABTCJQINMOMK9EDPnEZcDqDqJplijCq6QE7qeQaU4blSwM",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQCVnZ1On-Ja4xfAfMbsq--jatb5sNnOUN421AHaXbebcCWH",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
]
`;

exports[`Splitter should split if zero balance 1`] = `
[
  {
    "type": "deploy",
  },
  {
    "message": {
      "body": {
        "cell": "x{23D421E100000000000000E4}",
        "type": "cell",
      },
      "bounce": false,
      "from": "kQC3OBC-eWfUrnx1Yp6h6sOj779mNY3u3bK4idWfAdFVe7sB",
      "to": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
      "type": "internal",
      "value": 10000000000n,
    },
    "type": "received",
  },
  {
    "gasUsed": 77540n,
    "type": "processed",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQABTCJQINMOMK9EDPnEZcDqDqJplijCq6QE7qeQaU4blSwM",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/3 of 9.9 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQAQUQRWN2f1Dg5lycdymlpECWsWQg-WTvV8gVxnmcKgn2xQ",
        "to": "kQCVnZ1On-Ja4xfAfMbsq--jatb5sNnOUN421AHaXbebcCWH",
        "type": "internal",
        "value": 3298604000n,
      },
    ],
    "type": "sent",
  },
]
`;

exports[`Splitter should split revenue correctly 1`] = `
[
  {
    "type": "deploy",
  },
  {
    "message": {
      "body": {
        "text": "Gift",
        "type": "text",
      },
      "bounce": true,
      "from": "kQC3OBC-eWfUrnx1Yp6h6sOj779mNY3u3bK4idWfAdFVe7sB",
      "to": "kQBGGqhORfafzJLVcL2UaT5i6HzVLGCVt_dKIcEaA7oRcsnj",
      "type": "internal",
      "value": 2100000000n,
    },
    "type": "received",
  },
  {
    "gasUsed": 54339n,
    "type": "processed",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/2 of 2 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQBGGqhORfafzJLVcL2UaT5i6HzVLGCVt_dKIcEaA7oRcsnj",
        "to": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
        "type": "internal",
        "value": 998620000n,
      },
    ],
    "type": "sent",
  },
  {
    "messages": [
      {
        "body": {
          "text": "1/2 of 2 (whales revenue share)",
          "type": "text",
        },
        "bounce": false,
        "from": "kQBGGqhORfafzJLVcL2UaT5i6HzVLGCVt_dKIcEaA7oRcsnj",
        "to": "kQCVnZ1On-Ja4xfAfMbsq--jatb5sNnOUN421AHaXbebcCWH",
        "type": "internal",
        "value": 998620000n,
      },
    ],
    "type": "sent",
  },
]
`;
